docs/reference/syntax.md

```diff
@@ -0,0 +1,935 @@
+
+- Sections must be defined in the order shown above, with the exception of the output section, which can alternatively be specified after the script and stub.
+
+Each section may contain one or more statements. For directives, inputs, and outputs, these statements must be [function calls](#function-call). Refer to {ref}`process-reference` for the set of available input qualifiers, output qualifiers, and directives.
+
+The script section can be substituted with a shell or exec section:
+
+```groovy
+process greetShell {
+    input: 
+    val greeting
+
+    shell:
+    '''
+    echo '!{greeting}, ${USER}!'
+    '''
+}
+
+process greetExec {
+    input: 
+    val greeting
+    val name
+
+    exec:
+    message = "${greeting}, ${name}!"
+
+    output:
+    val message
+}
+```
+
+The script, shell, and stub sections must return a string in the same manner as a [function](#function).
+
+Refer to {ref}`process-page` for more information on the semantics of each process section.
+
+(syntax-function)=
+
+### Function
+
+A function consists of a name, parameter list, and a body:
+
+```groovy
+def greet(greeting, name) {
+    println "${greeting}, ${name}!"
+}
+```
+
+The function body consists of one or more [statements](#statements). The last statement is implicitly treated as a return statement if it is an [expression statement](#expression-statement) that returns a value.
+
+The [return statement](#return) can be used to explicitly return from a function:
+
+```groovy
+// return with no value
+def greet(greeting, name) {
+    if( !greeting || !name )
+        return
+    println "${greeting}, ${name}!"
+}
+
+// return a value
+def fib(x) {
+    if( x <= 1 )
+        return x
+    fib(x - 1) + fib(x - 2)
+}
+```
+
+### Enum type
+
+An enum type declaration consists of a name and a body, which consists of a comma-separated list of identifiers:
+
+```groovy
+enum Day {
+    MONDAY,
+    TUESDAY,
+    WEDNESDAY,
+    THURSDAY,
+    FRIDAY,
+    SATURDAY,
+    SUNDAY
+}
+```
+
+Enum values can be accessed as `Day.MONDAY`, `Day.TUESDAY`, and so on.
+
+:::{note}
+Enum types cannot be included across modules at this time.
+:::
+
+### Output block
+
+The output block consists of one or more *target blocks*. A target block consists of a *target name* and one or more *target directives* for configuring the corresponding publish target:
+
+```groovy
+output {
+    'fastq' {
+        path 'samples'
+        index {
+            path 'index.csv'
+        }
+    }
+}
+```
+
+Only one output block may be defined in a script. Refer to {ref}`workflow-output-def` for the set of available target directives.
+
+## Statements
+
+Statements should be separated by a newline or semi-colon:
+
+```groovy
+// newline
+println 'Hello!'
+println 'Hello again!'
+
+// semi-colon
+println 'Hello!' ; println 'Hello again!'
+```
+
+### Variable declaration
+
+Variables can be declared with the `def` keyword:
```

Comment:
I am stilling thinking about the best way to communicate this, because for a time there will be a mis-match between the formal syntax and what Nextflow allows in practice

From a "formal syntax" point of view, I think we should simply say that `def` is always required. But in practice there is still a good reason to omit `def` in a process (so that you can re-use the variable in the outputs). But it seems weird to talk about processes in this section about variable declarations

So for now there is a note in the Scripts page to the effect of "sometimes variables can be declared with `def` but this is usually dangerous so you shouldn't do it"

Reply:
I think the Scripts page is the appropriate place to address this for now, as it deals more with semantics (i.e. meaning) while this page is focused on syntax (i.e. structure)

Both of these pages can still be expanded significantly, but I'd rather save that for a future iteration